### PR TITLE
Internal publishing service

### PR DESCRIPTION
**Pull request by mvandenburgh**
Refactors publishing into a service layer module.

### CODE REVIEW

**Comment by mvandenburgh** [comment]
whoops I forgot to check if tests were passing before opening this...

File: dandiapi/api/services/publish/__init__.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+def publish_dandiset(*, user: User, dandiset: Dandiset) -> None:
+    from dandiapi.api.tasks import publish_dandiset_task
+
+    if (
+        not user.has_perm('owner', dandiset)
+        or dandiset.embargo_status != Dandiset.EmbargoStatus.OPEN
+    ):
+        raise NotAllowed()
+    if dandiset.zarr_archives.exists() or dandiset.embargoed_zarr_archives.exists():
+        raise NotAllowed('Cannot publish dandisets which contain zarrs.')
```
Comment (AlmightyYakob) [inline dandiapi/api/services/publish/__init__.py:132]: It seems this exact code block exists at the beginning of `_lock_dandiset_for_publishing`, which immediately follows here. So it seems this check is redundant? I'd understand if it had to stay to preserve order of operations, but it seems that wouldn't be impacted by its removal.

**Review by AlmightyYakob — CHANGES_REQUESTED** [review]
One comment but overall looks good. The structure does bug me a bit (bouncing between the publish service and tasks files), but I'm not sure of a good way to address that.

File: dandiapi/api/tests/test_version.py [hunk 2]
```diff
@@ -654,8 +654,8 @@ def test_version_rest_publish_zarr(
         f'/api/dandisets/{draft_version.dandiset.identifier}'
         f'/versions/{draft_version.version}/publish/'
     )
-    assert resp.status_code == 400
-    assert resp.json() == ['Cannot publish dandisets which contain zarrs']
+    assert resp.status_code == 403
```
Comment (danlamanna) [inline dandiapi/api/tests/test_version.py:657]: Just a note to back out this one change we discussed.

Comment (mvandenburgh) [reply dandiapi/api/tests/test_version.py]: :+1:

File: dandiapi/api/services/publish/__init__.py (hunk 1 above)
Comment (mvandenburgh) [reply dandiapi/api/services/publish/__init__.py]: [`ac75bf5` (#1363)](https://github.com/dandi/dandi-archive/pull/1363/commits/ac75bf5dd61958debc0ac32d50e4be81b39be850)

**Comment by mvandenburgh** [comment]
> The structure does bug me a bit (bouncing between the publish service and tasks files), but I'm not sure of a good way to address that.

Yeah I'm still not completely happy with how that turned out :(
Definitely open to suggestions if you think of any.

**Review by AlmightyYakob — APPROVED** [review]

**Pull request merged**